accord-core/src/test/java/accord/impl/basic/Cluster.java

```diff
@@ -301,4 +345,147 @@ public static EnumMap<MessageType, Stats> run(Id[] nodes, Supplier<PendingQueue>
             lookup.values().forEach(Node::shutdown);
         }
     }
+
+    private static BiFunction<Id, Id, Link> partition(List<Id> nodes, RandomSource random, int rf, BiFunction<Id, Id, Link> defaultLinks)
+    {
+        Collections.shuffle(nodes, random.asJdkRandom());
+        int partitionSize = random.nextInt((rf+1)/2);
+        Set<Id> partition = new LinkedHashSet<>(nodes.subList(0, partitionSize));
+        return (from, to) -> partition.contains(from) == partition.contains(to)
+                             ? defaultLinks.apply(from, to)
+                             : Link.DOWN;
+    }
+
+    /**
+     * pair every node with one other node in one direction with a network behaviour override
+     */
+    private static BiFunction<Id, Id, Link> pairedUnidirectionalOverrides(Function<Link, Link> linkOverride, List<Id> nodes, RandomSource random, BiFunction<Id, Id, Link> fallback)
+    {
+        Map<Id, Map<Id, Link>> map = new HashMap<>();
+        Collections.shuffle(nodes, random.asJdkRandom());
```

Comment:
```suggestion
        List<Id> nodes = new ArrayList<>(nodes);
        Collections.shuffle(nodes, random.asJdkRandom());
```

mutating a list given to you can be a source of bugs... would recommend moving away from that and treat your input as immutable...

Reply:
We've already done this outside of this inside of `run` prior to passing it into these methods.

Reply:
We take a copy of the list externally to the generators, and invoke them in a single thread. So I don't think this is necessary, but if you strongly prefer it I don't mind applying a local copy as well.

Reply:
ill leave this as a +0... tend to find coupling like this change over time and cause subtle issues so I prefer to be more defensive, but if you know the ownership within the whole context is correct, we can leave as is.